log4j-api/src/main/java/org/apache/logging/log4j/status/StatusLogger.java

```diff
@@ -16,311 +16,567 @@
+        buffer(statusData);
+        notifyListeners(statusData);
+    }
+
+    private void buffer(final StatusData statusData) {
+        if (config.bufferCapacity == 0) {
+            return;
         }
-        final StatusData data = new StatusData(element, level, msg, t, null);
-        msgLock.lock();
-        try {
-            messages.add(data);
-        } finally {
-            msgLock.unlock();
+        buffer.add(statusData);
+        while (buffer.size() >= config.bufferCapacity) {
+            buffer.remove();
         }
-        // LOG4J2-1813 if system property "log4j2.debug" is defined, all status logging is enabled
-        if (DEBUG_ENABLED || (listeners.size() <= 0)) {
-            logger.logMessage(fqcn, level, marker, msg, t);
-        } else {
-            for (final StatusListener listener : listeners) {
-                if (data.getLevel().isMoreSpecificThan(listener.getStatusLevel())) {
-                    listener.log(data);
+    }
+
+    private void notifyListeners(final StatusData statusData) {
+        listenerReadLock.lock();
+        try {
+            listeners.forEach(listener -> {
+                if (config.debugEnabled || listener.getStatusLevel().isLessSpecificThan(statusData.getLevel())) {
+                    listener.log(statusData);
                 }
-            }
+            });
+        } finally {
+            listenerReadLock.unlock();
         }
     }
 
-    private StackTraceElement getStackTraceElement(final String fqcn, final StackTraceElement[] stackTrace) {
+    private StatusData createStatusData(
+            @Nullable final String fqcn,
+            final Level level,
+            final Message message,
+            @Nullable final Throwable throwable) {
+        final StackTraceElement caller = getStackTraceElement(fqcn);
+        return new StatusData(caller, level, message, throwable, null, config.instantFormatter);
+    }
+
+    @Nullable
+    private static StackTraceElement getStackTraceElement(@Nullable final String fqcn) {
         if (fqcn == null) {
             return null;
         }
         boolean next = false;
+        final StackTraceElement[] stackTrace = Thread.currentThread().getStackTrace();
         for (final StackTraceElement element : stackTrace) {
             final String className = element.getClassName();
             if (next && !fqcn.equals(className)) {
                 return element;
             }
             if (fqcn.equals(className)) {
                 next = true;
-            } else if (NOT_AVAIL.equals(className)) {
+            } else if ("?".equals(className)) {
                 break;
             }
         }
         return null;
     }
 
     @Override
-    public boolean isEnabled(final Level level, final Marker marker, final String message, final Throwable t) {
+    public boolean isEnabled(final Level level, final Marker marker, final String message, final Throwable throwable) {
         return isEnabled(level, marker);
     }
```

Comment:
It is probably outside of the scope of this PR, but we implement these `ExtendedLogger` methods in the same way multiple times in our codebase.

This should be a `default` method on the `ExtendedLogger`, since it works for all implementations except `log4j-core`. We could even add:

```java
default boolean isEnabled(final Level level, final Marker marker) {
    return isEnabled(level);
}
```
since this works for all implementations except `log4j-to-slf4j` (and obviously `log4j-core`).

Reply:
We can indeed enhance `ExtendedLogger` with `default`s. But I prefer to do it outside this PR.